Desktop_Interface/isobuffer.cpp

```diff
@@ -2,391 +2,364 @@
-    samplesPerSecond = samplesPerSecond/375*VALID_DATA_PER_375;
-    sampleRate_bit = samplesPerSecond * 8;
-    virtualParent = caller;
-    channel = channel_value;
+	buffer = (short*) calloc(bufferLen*2, sizeof(short));
+	bufferEnd = bufferLen-1;
+	samplesPerSecond = (double) bufferLen/(double)21;
+	samplesPerSecond = samplesPerSecond/375*VALID_DATA_PER_375;
+	sampleRate_bit = samplesPerSecond * 8;
+	virtualParent = caller;
+	channel = channel_value;
 }
+
 void isoBuffer::openFile(QString newFile)
 {
-    if (fptr != NULL){
-        fclose(fptr);
-    }
-    if (newFile.isEmpty()){
-        fptr = NULL;
-    }
-    else {
-        QByteArray temp = newFile.toLatin1();
-        char *fileName = temp.data();
-        fptr = fopen(fileName, "w");
-        if (fptr == NULL) qFatal("Null fptr in isoBuffer::openFile");
-        qDebug() << "opening file" << fileName;
-        qDebug() << "fptr = " << fptr;
-    }
+	if (fptr != NULL)
+	{
+		fclose(fptr);
+	}
+	if (newFile.isEmpty())
+	{
+		fptr = NULL;
+	}
+	else
+	{
+		QByteArray temp = newFile.toLatin1();
+		char* fileName = temp.data();
+		fptr = fopen(fileName, "w");
+		if (fptr == NULL) qFatal("Null fptr in isoBuffer::openFile");
+		qDebug() << "opening file" << fileName;
+		qDebug() << "fptr = " << fptr;
+	}
+}
+
+void isoBuffer::insertIntoBuffer(short item)
+{
+	buffer[back] = item;
+	back++;
+	insertedCount++;
+	if (insertedCount > bufferEnd)
+	{
+		insertedCount = bufferEnd+1;
+	}
+	if (back > bufferEnd)
+	{
+		back = 0;
+	}
+}
+
+short isoBuffer::bufferAt(int idx) const
+{
+	return buffer[back - idx];
+}
+
+bool isoBuffer::maybeOutputSampleToFile(double convertedSample)
+{
+	/*
+	 * This function adds a sample to an accumulator and bumps the sample count
+	 * after the sample count hits some threshold, that sample is outputted to a
+	 * file. If this 'saturates' the file, then fileIO is disabled.
+	 */
+	average_sample_temp += convertedSample;
+	fileIO_sampleCount++;
+
+	//Check to see if we can write a new sample to file
+	if (fileIO_sampleCount == fileIO_maxIncrementedSampleValue)
+	{
+		char numStr[32];
+		sprintf(numStr,"%7.5f, ", average_sample_temp/((double)fileIO_maxIncrementedSampleValue));
+		currentFile->write(numStr);
+		currentColumn++;
+		if (currentColumn >= COLUMN_BREAK)
+		{
+			currentFile->write("\n");
+			currentColumn = 0;
+		}
+
+		//Reset the average and sample count for next data point
+		fileIO_sampleCount = 0;
+		average_sample_temp = 0;
+
+		//Check to see if we've reached the max file size.
+		if (fileIO_max_file_size != 0) //value of 0 means "no limit"
+		{
+			fileIO_numBytesWritten += 9;  //7 chars for the number, 1 for the comma and 1 for the space = 9 bytes per sample.
+			if (fileIO_numBytesWritten >= fileIO_max_file_size)
+			{
+				fileIOEnabled = false; //Just in case signalling fails.
+				fileIOinternalDisable();
+				return false;
+			}
+		}
+	}
+	return true;
 }
 
 void isoBuffer::writeBuffer_char(char* data, int len)
```

Comment:
Wow.  Good work refactoring these.  Just a little bit more concise now!
Could template use allow us to roll them into a single function now, though, since they share exactly the same source code?

Reply:
Oh yes. I was thinking about that but was unsure about what your policy on templates was, so I didn't want to try that just yet.

Reply:
Also: I happened to look through isoBuffer_file and it looks like it should eventually be possible to extract the common functionality of isoBuffer and isoBuffer_file to a separate class :)